### PR TITLE
IMPRO-1643: Make use of BasePlugin.process calls

### PR DESCRIPTION
**Pull request by bayliffe**
This PR modifies all IMPROVER code (CLIs/plugins/functions) to make use of the BasePlugin.process method that makes our plugins callable without reference to the .process method.

I have removed the merge_cubes wrapper to MergeCubes as the latter can now be called in much the same way as the wrapper was.

I have not updated unit tests, which still call the .process method directly when testing a plugin. 

There are a few small pylint fixes that were made whilst modifying files (e.g. grouping iris imports).

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver/cli/weighted_blending.py [hunk 1]
```diff
@@ -141,12 +141,11 @@ def process(*cubes: cli.inputcube,
...
         weighting_coord=weighting_coord, wts_dict=weighting_config,
-        y0val=y0val, ynval=ynval, cval=cval)
-    result = plugin.process(
-        cubes, cycletime=cycletime, model_id_attr=model_id_attr,
-        spatial_weights=spatial_weights_from_mask, fuzzy_length=fuzzy_length,
-        attributes_dict=attributes_config)
+        y0val=y0val, ynval=ynval, cval=cval)(
+            cubes, cycletime=cycletime, model_id_attr=model_id_attr,
+            spatial_weights=spatial_weights_from_mask,
+            fuzzy_length=fuzzy_length, attributes_dict=attributes_config)
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/cli/weighted_blending.py:150]: I don't have strong feelings either way. I was endeavouring to make all the ones I touched consistent in style. The differing indentation levels help show the distinction as things stand, but if the 2nd reviewer agrees that splitting these back up is a good idea, I will do so.

File: improver/cli/apply_emos_coefficients.py [hunk 2]
```diff
@@ -214,25 +214,25 @@ def process(cube: cli.inputcube,
...
         percentiles = ConvertLocationAndScaleParametersToPercentiles(
             distribution=distribution,
-            shape_parameters=shape_parameters).process(
-            location_parameter, scale_parameter, original_current_forecast,
-            no_of_percentiles=no_of_percentiles)
-        result = EnsembleReordering().process(
+            shape_parameters=shape_parameters)(
+                location_parameter, scale_parameter, original_current_forecast,
+                no_of_percentiles=no_of_percentiles)
+        result = EnsembleReordering()(
```
Comment (cgsandford) [inline improver/cli/apply_emos_coefficients.py:235]: I agree with Fiona that some of the calls with very long lists of "process" arguments would be clearer separated.  So I'll note that on that comment.  But in this particular file (apply EMOS CLI) may I ask you just to leave it as you have, as I'm currently refactoring this anyway so will do some cleanup then.

File: improver/cli/estimate_emos_coefficients.py [hunk 3]
```diff
@@ -112,5 +112,5 @@ def process(*cubes: cli.inputcube,
     return EstimateCoefficientsForEnsembleCalibration(
         distribution, cycletime, desired_units=units,
         predictor=predictor, tolerance=tolerance,
-        max_iterations=max_iterations).process(
+        max_iterations=max_iterations)(
```
Comment (cgsandford) [inline improver/cli/estimate_emos_coefficients.py:115]: Long init list - could you split this into `plugin =  ...` and `return plugin.process(...` please.

File: improver/cli/generate_orographic_smoothing_coefficients.py [hunk 4]
```diff
@@ -86,4 +86,4 @@ def process(orography: cli.inputcube,
         OrographicSmoothingCoefficients)
     return OrographicSmoothingCoefficients(
         min_smoothing_coefficient, max_smoothing_coefficient, coefficient,
-        power).process(orography)
+        power)(orography)
```
Comment (cgsandford) [inline improver/cli/generate_orographic_smoothing_coefficients.py:89]: Long init list...

File: improver/cli/neighbour_finding.py [hunk 5]
```diff
@@ -186,16 +186,16 @@ def process(orography: cli.inputcube,
...
-            all_methods.append(NeighbourSelection(**method).process(*fargs))
+            all_methods.append(NeighbourSelection(**method)(*fargs))
 
         squeezed_cubes = iris.cube.CubeList([])
         for index, cube in enumerate(all_methods):
             cube.coord('neighbour_selection_method').points = np.int32(index)
             squeezed_cubes.append(iris.util.squeeze(cube))
 
-        result = merge_cubes(squeezed_cubes)
+        result = MergeCubes()(squeezed_cubes)
```
Comment (cgsandford) [inline improver/cli/neighbour_finding.py:196]: Have you got rid of all of the `merge_cubes` calls?  Because if so that wrapper function can be removed.  (Don't spend extra time checking if you didn't intend this, though.)

File: improver/cli/nowcast_accumulate.py [hunk 6]
```diff
@@ -103,8 +103,8 @@ def process(cube: cli.inputcube,
     # extrapolate input data to the maximum required lead time
     forecast_cubes = CreateExtrapolationForecast(
         cube, u_cube, v_cube, orographic_enhancement,
-        attributes_dict=attributes_config).process(ACCUMULATION_FIDELITY,
-                                                   max_lead_time)
+        attributes_dict=attributes_config)(ACCUMULATION_FIDELITY,
+                                           max_lead_time)
```
Comment (cgsandford) [inline improver/cli/nowcast_accumulate.py:107]: Long init list

File: improver/cli/nowcast_accumulate.py [hunk 7]
```diff
@@ -113,6 +113,6 @@ def process(cube: cli.inputcube,
     result = Accumulation(
         accumulation_units=accumulation_units,
         accumulation_period=accumulation_period * 60,
-        forecast_periods=lead_times * 60).process(forecast_cubes)
+        forecast_periods=lead_times * 60)(forecast_cubes)
```
Comment (cgsandford) [inline improver/cli/nowcast_accumulate.py:116]: Long init list

File: improver/cli/nowcast_extrapolate.py [hunk 8]
```diff
@@ -73,14 +73,13 @@ def process(cube: cli.inputcube,
...
 
     u_cube, v_cube = advection_velocity
 
     # extrapolate input data to required lead times
-    forecast_plugin = CreateExtrapolationForecast(
+    forecast_cubes = CreateExtrapolationForecast(
         cube, u_cube, v_cube, orographic_enhancement,
-        attributes_dict=attributes_config)
-    forecast_cubes = forecast_plugin.process(lead_time_interval, max_lead_time)
+        attributes_dict=attributes_config)(lead_time_interval, max_lead_time)
```
Comment (cgsandford) [inline improver/cli/nowcast_extrapolate.py:83]: Long init list

File: improver/cli/standardise.py [hunk 9]
```diff
@@ -133,13 +133,12 @@ def process(cube: cli.inputcube,
...
         landmask=land_sea_mask,
-        landmask_vicinity=land_sea_mask_vicinity)
-    output_data = plugin.process(
-        cube, target_grid, new_name=new_name, new_units=new_units,
-        regridded_title=regridded_title, coords_to_remove=coords_to_remove,
-        attributes_dict=attributes_config)
+        landmask_vicinity=land_sea_mask_vicinity)(
+            cube, target_grid, new_name=new_name, new_units=new_units,
+            regridded_title=regridded_title, coords_to_remove=coords_to_remove,
+            attributes_dict=attributes_config)
```
Comment (cgsandford) [inline improver/cli/standardise.py:142]: Definitely split this one back up!

File: improver/cli/threshold.py [hunk 10]
```diff
@@ -159,5 +159,5 @@ def process(cube: cli.inputcube,
         warnings.warn("Collapse-coord option not fully tested with "
                       "masked data.")
     # Take a weighted mean across realizations with equal weights
-    plugin = WeightAndBlend(collapse_coord, "linear", y0val=1.0, ynval=1.0)
-    return plugin.process(result_no_collapse_coord)
+    return WeightAndBlend(collapse_coord, "linear", y0val=1.0, ynval=1.0)(
+        result_no_collapse_coord)
```
Comment (cgsandford) [inline improver/cli/threshold.py:163]: Yes this would be clearer as two statements.

File: improver/cli/weighted_blending.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/cli/weighted_blending.py:150]: I agree with Fiona on this.  I think consistency of calls is a good thing but probably not the important thing in this case - since all the plugins are so different in how they look and function (number of arguments, init / process split, etc), aiming for consistency probably actually makes the code base less clear.  We might be able to aim for that in the future though.

File: improver/cli/wind_direction.py [hunk 11]
```diff
@@ -65,5 +65,5 @@ def process(wind_direction: cli.inputcube,
     from improver.wind_calculations.wind_direction import WindDirection
 
     result, _, _ = (
-        WindDirection(backup_method=backup_method).process(wind_direction))
+        WindDirection(backup_method=backup_method)(wind_direction))
```
Comment (cgsandford) [inline improver/cli/wind_direction.py:68]: Is there a better way to phrase this call, or is this "best practise" style?  (Eg `result, = WindDirection(...` or `result, ... = WindDirection(...`?)  This just looks horrible to me.

File: improver/ensemble_copula_coupling/ensemble_copula_coupling.py [hunk 12]
```diff
@@ -1248,7 +1248,7 @@ def process(
             random_ordering=random_ordering,
             random_seed=random_seed)
         post_processed_forecast_realizations = (
-            RebadgePercentilesAsRealizations.process(
+            RebadgePercentilesAsRealizations()(
                 post_processed_forecast_realizations))
```
Comment (cgsandford) [inline improver/ensemble_copula_coupling/ensemble_copula_coupling.py:1252]: Soooo many brackets...

File: improver/nbhood/use_nbhood.py [hunk 13]
```diff
@@ -192,7 +192,7 @@ def process(self, cube, mask_cube):
                     lead_times=self.lead_times,
                     weighted_mode=self.weighted_mode,
                     sum_or_fraction=self.sum_or_fraction, re_mask=self.re_mask
-                    ).process(x_y_slice, mask_cube=cube_slice)
+                    )(x_y_slice, mask_cube=cube_slice)
```
Comment (cgsandford) [inline improver/nbhood/use_nbhood.py:195]: Long init list

File: improver/orographic_enhancement.py [hunk 14]
```diff
@@ -270,7 +270,7 @@ def _generate_mask(self):
         # calculate mean 3x3 (square nbhood) orography heights
         radius = number_of_grid_cells_to_distance(self.topography, 1)
         topo_nbhood = NeighbourhoodProcessing(
-            'square', radius).process(self.topography)
+            'square', radius)(self.topography)
```
Comment (cgsandford) [inline improver/orographic_enhancement.py:273]: Does this line unwrap?

File: improver_tests/standardise/test_AdjustLandSeaPoints.py [hunk 15]
```diff
@@ -63,7 +63,7 @@ def test_basic(self):
                             'output_cube': None}
         result = AdjustLandSeaPoints()
         members = {attr: getattr(result, attr) for attr in dir(result)
-                   if not attr.startswith("__")}
+                   if not attr.startswith("_")}
```
Comment (cgsandford) [inline improver_tests/standardise/test_AdjustLandSeaPoints.py:66]: Why this change?

File: improver/cli/neighbour_finding.py (hunk 5 above)
Comment (cgsandford) [reply improver/cli/neighbour_finding.py:196]: You're way ahead of me!  Thanks for this tidy-up.

File: improver_tests/standardise/test_AdjustLandSeaPoints.py (hunk 15 above)
Comment (bayliffe) [reply improver_tests/standardise/test_AdjustLandSeaPoints.py:66]: Because the AdjustLandSeaPoints class now inherits the base class attributes, some of which begin with a single ```_```. So if they are not excluded here, the test will fail. Previously no single underscore attributes existed for this class.

File: improver/cli/apply_emos_coefficients.py (hunk 2 above)
Comment (bayliffe) [reply improver/cli/apply_emos_coefficients.py:235]: Leaving alone.

File: improver/cli/estimate_emos_coefficients.py (hunk 3 above)
Comment (bayliffe) [reply improver/cli/estimate_emos_coefficients.py:115]: Done (without the .process obviously).

File: improver/cli/generate_orographic_smoothing_coefficients.py (hunk 4 above)
Comment (bayliffe) [reply improver/cli/generate_orographic_smoothing_coefficients.py:89]: Done.

File: improver/cli/nowcast_accumulate.py (hunk 6 above)
Comment (bayliffe) [reply improver/cli/nowcast_accumulate.py:107]: Done.

File: improver/cli/nowcast_accumulate.py (hunk 7 above)
Comment (bayliffe) [reply improver/cli/nowcast_accumulate.py:116]: Done.

File: improver/cli/nowcast_extrapolate.py (hunk 8 above)
Comment (bayliffe) [reply improver/cli/nowcast_extrapolate.py:83]: Done.

File: improver/cli/standardise.py (hunk 9 above)
Comment (bayliffe) [reply improver/cli/standardise.py:142]: Done.

File: improver/cli/threshold.py (hunk 10 above)
Comment (bayliffe) [reply improver/cli/threshold.py:163]: Done.

File: improver/cli/wind_direction.py (hunk 11 above)
Comment (bayliffe) [reply improver/cli/wind_direction.py:68]: I've stuck a [0] index on the end of the call which means we simply don't assign the [1] and [2] elements rather than assigning them to null. Is this preferred? It is perhaps a little more hidden with that little [0] tucked at the end.

File: improver/ensemble_copula_coupling/ensemble_copula_coupling.py (hunk 12 above)
Comment (bayliffe) [reply improver/ensemble_copula_coupling/ensemble_copula_coupling.py:1252]: Broken up.

File: improver/nbhood/use_nbhood.py (hunk 13 above)
Comment (bayliffe) [reply improver/nbhood/use_nbhood.py:195]: Took the initialisation out of the loop.

File: improver/cli/wind_direction.py (hunk 11 above)
Comment (fionaRust) [reply improver/cli/wind_direction.py]: I thought the underscore way of doing it was a standard python thing, but I might be wrong.
